### PR TITLE
Motion detect

### PR DESCRIPTION
**Pull request by mzgoddard**
### Resolves

Incremental part of #304

### Proposed Changes

- Add rough version of videoSensing extension
- Implement scratch 2 `senseVideoMotion` opcode as `videoSensing.videoOn`
- Add a debug and benchmark page for video sensing

### Test Coverage

Part of a future PR.

-----

This is an incremental implementation of the Video Sensing extension. It implements the the video motion input block but does not yet implement the video motion event block. The video motion block supports scratch 2's motion and direction attributes of either the stage or 'this sprite'.

The extension started as a clone of the music extension so there may remain some parts in the video sensing extension that may not be needed. As an incremental step to implement video sensing, we could merge this in while further PRs I'm working on add the event block and do further documentation, clean up, and testing. Review on things like names displayed in the gui or argument names would be great to get.

### CODE REVIEW

File: src/extensions/scratch3_video_sensing/index.js [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            this._drawable,
+            Scratch3VideoSensingBlocks.ORDER
+        );
+        this.runtime.renderer.updateDrawableProperties(this._drawable, {
+            skinId: this._skinId
+        });
+    }
+
+    _loop () {
+        setTimeout(this._loop.bind(this), this.runtime.THREAD_STEP_INTERVAL_COMPATIBILITY);
```
Comment (cwillisf) [inline src/extensions/scratch3_video_sensing/index.js]: Should this use `Scratch3VideoSensingBlocks.INTERVAL` instead? Or perhaps it should be `runtime.currentStepTime`?

File: src/extensions/scratch3_video_sensing/lib.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        let A2;
+        let A1B2;
+        let B1;
+        let C1;
+        let C2;
+        let u;
+        let v;
+        let uu;
+        let vv;
+        let n;
```
Comment (cwillisf) [inline src/extensions/scratch3_video_sensing/lib.js]: I'd suggest moving these inside the relevant scopes, where possible. For example, it looks like the `A`/`B`/`C` family could move into the `j` loop, to where they get initialized each loop. Unless JS is silly enough to cause a performance hit in a case like that...?

File: src/extensions/scratch3_video_sensing/view.js [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+        this.output = output;
+        this.buffer = new Uint32Array(WIDTH * HEIGHT);
+    }
+
+    static get OUTPUT () {
+        return OUTPUT;
+    }
+
+    draw () {
```
Comment (cwillisf) [inline src/extensions/scratch3_video_sensing/view.js]: There's a lot of code duplicated here... is that intentional? Could they share code, or is one of the copies obsolete?

File: webpack.config.js [hunk 4]
```diff
@@ -86,7 +86,8 @@ module.exports = [
                 'scratch-storage',
                 // Renderer
                 'scratch-render'
-            ]
+            ],
+            'motion-extension': './src/extensions/scratch3_video_sensing/debug'
```
Comment (cwillisf) [inline webpack.config.js:90]: It shouldn't be necessary to explicitly mention the sensing extension in the webpack config as long as it's `require`d as you do in `extension-manager.js`. If you want to expose debug stuff to the JS console, I would use `global` (which Webpack turns into `self` or `window`). I imagine you're familiar, but if not, you can find an example in `src/extension-support/extension-worker.js`.

**Review by cwillisf — CHANGES_REQUESTED** [review]
I have a few nitpicks but it looks like you're making great progress! 📹

File: src/extensions/scratch3_video_sensing/index.js (hunk 1 above)
Comment (thisandagain) [reply src/extensions/scratch3_video_sensing/index.js]: 👍 That is actually my bad from the Vision extension. Sorry @mzgoddard !

File: src/extensions/scratch3_video_sensing/source.js [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (thisandagain) [inline src/extensions/scratch3_video_sensing/source.js]: Can this be removed?

File: webpack.config.js (hunk 4 above)
Comment (mzgoddard) [reply webpack.config.js:90]: This is used by `src/playground/motion.html`.

File: src/extensions/scratch3_video_sensing/source.js (hunk 5 above)
Comment (mzgoddard) [reply src/extensions/scratch3_video_sensing/source.js]: Yes.

File: src/extensions/scratch3_video_sensing/index.js (hunk 1 above)
Comment (mzgoddard) [reply src/extensions/scratch3_video_sensing/index.js]: That's a good point @cwillisf. Eventually we probably want to use [`timeupdate`](https://developer.mozilla.org/en-US/docs/Web/Events/timeupdate) or some thing to execute a step of the loop approximately at the hertz of the camera.

File: src/extensions/scratch3_video_sensing/lib.js (hunk 2 above)
Comment (mzgoddard) [reply src/extensions/scratch3_video_sensing/lib.js]: I think that'll be better for readability.

File: src/extensions/scratch3_video_sensing/view.js (hunk 3 above)
Comment (mzgoddard) [reply src/extensions/scratch3_video_sensing/view.js]: Its duplicated intentionally. A lot of it was finding what needed to be copied to get the output renders to help understand the algorithm output in real time. The gradX, gradY, gradT calculations could be deduped. Some of the copies are out of date in minor ways.

File: src/extensions/scratch3_video_sensing/index.js [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        return 'Scratch.videoSensing';
+    }
+
+    /**
+     * The default music-related state, to be used when a target has no existing music state.
+     * @type {MusicState}
+     */
+    static get DEFAULT_MOTION_STATE () {
+        return {
+            currentInstrument: 0
```
Comment (gnarf) [inline src/extensions/scratch3_video_sensing/index.js:197]: ?? copy/pasta leftover from music?

**Comment by mzgoddard** [comment]
@cwillisf Do you see any more improvements this needs?

**Review by cwillisf — APPROVED** [review]
Looks good, assuming that future work will clean up the remaining commented and/or music-related bits.

**Pull request merged**